File: Programmers/Lv3/Lv3_가장먼노드.cpp
#include <string>
#include <vector>
#include <queue>
#include <list>
#include <algorithm>
#include <limits>
using namespace std;

// 다익스트라 알고리즘보다 BFS가 훨씬 빠름.
// 가중치가 1일 때는 BFS를 이용하자!

int solution(int n, vector<vector<int>> edge) {
	int answer = 0;
	int inf = 0;
	vector<list<int> > graph(n+1, list<int>());
	vector<int> dist(n+1, 0);
	vector<bool> check(n+1, false);
	queue<int> q;

	for (auto i = edge.begin(); i != edge.end(); i++) {		// 1-2, 2-1 두 개로 저장
		graph[(*i).front()].push_back((*i).back());
		graph[(*i).back()].push_back((*i).front());
	}

	q.push(1);
	check[1] = true;
	dist[1] = 0;

	while (!q.empty()) {
		int start = q.front(); 
		q.pop();
		for (auto i : graph[start]) {
			if (check[i]==false) {
				check[i] = true;
				dist[i] = dist[start] + 1;
				q.push(i);
			}
		}
	}

	sort(dist.begin(), dist.end());
	inf = dist.back();

	for (int i = 1; i <= n;  i++) {
		if (dist[i] == inf)
			answer++;
	}
	return answer;
}

File: Programmers/Lv3/Lv3_가장먼노드.py
def solution(n, edge):
    answer = 0
    graph = [set() for x in range(n+1)]
    dist = [0] * (n+1)
    check = [False] * (n+1)
    q = []

    for i in edge:                      # ?곌껐 ?뺣낫 ???
        graph[i[0]].add(i[1])
        graph[i[1]].add(i[0])

    q.append(1)
    check[1] = True
    dist[1] = 0

    while q:
        start = q.pop(0)
        for i in graph[start]:
            if not check[i]:
                check[i] = True
                dist[i] = dist[start] + 1
                q.append(i)

    inf = max(dist)                 # 媛??癒?嫄곕━
    for i in range(1, n+1):
        if dist[i] == inf:
            answer += 1
    return answer

print(solution(6, [[3, 6], [4, 3], [3, 2], [1, 3], [1, 2], [2, 4], [5, 2]]))        # 3